components/log/log.c

```diff
@@ -0,0 +1,159 @@
+#include "log.h"
+
+#include <errno.h>
+#include <string.h>
+#include <unistd.h>
+
+// The names are long, so let's add a convenience helpers
+typedef datadog_php_logger logger_t;
+typedef datadog_php_log_level log_level_t;
+typedef datadog_php_string_view string_view_t;
+
+logger_t datadog_php_logger_ctor(int descriptor, log_level_t log_level, pthread_mutex_t *mutex) {
+    logger_t logger = {
+        .descriptor = descriptor < 0 ? -1 : descriptor,  // normalize invalid
+        .log_level = log_level,
+        .mutex = mutex,
+    };
+    return logger;
+}
+
+static bool logger_valid(datadog_php_logger *logger) {
+    return logger->descriptor >= 0 && logger->log_level >= DATADOG_PHP_LOG_OFF &&
+           logger->log_level <= DATADOG_PHP_LOG_DEBUG && logger->mutex;
+}
+
+bool datadog_php_logger_valid(datadog_php_logger *logger) {
+    bool valid = false;
+    if (logger->mutex && pthread_mutex_lock(logger->mutex) == 0) {
+        valid = logger_valid(logger);
+        (void)pthread_mutex_unlock(logger->mutex);
+    }
+    return valid;
+}
+
+void datadog_php_logger_dtor(logger_t *logger) {
+    logger->descriptor = -1;
+    logger->log_level = DATADOG_PHP_LOG_UNKNOWN;
+    logger->mutex = NULL;
+}
+
+static int64_t durable_write(const logger_t *logger, string_view_t message) {
+    /* We attempt to handle some of the error conditions possible:
+     *  1. Partial writes
+     *  2. EAGAIN and EWOULDBLOCK
+     *  3. EINTR
+     * We limit the number of times we'll retry an operation for the above.
+     * For other errors we stop attempting to log and move on.
+     * todo: should we set the fd to -1 for some errors like EIO/EPIPE?
+     */
+    size_t written = 0;
+    unsigned tries = 0, max_tries = 3;  // arbitrarily chosen
+    while (written < message.len && tries++ < max_tries) {
+        ssize_t n = write(logger->descriptor, message.ptr + written, message.len - written);
+        if (n < 0) {
+            /* POSIX.1-2001 allows either error to be returned for this case, and does
+             * not require these constants to have the same value, so a portable
+             * application should check for both possibilities.
+             */
+            if (errno == EAGAIN || errno == EWOULDBLOCK || errno == EINTR) {
```

Comment:
EWOULDBLOCK? Is the logging descriptor supposed to be non-blocking?

Reply:
The comment tries to explain that `EAGAIN` and `EWOULDBLOCK` are actually the same constant on many platforms, but this isn't guaranteed, so check for both. Since this is doing `write` on a potentially arbitrary socket, we just handle things in a basic way with immediate retry mechanism. There are certainly cases it will not handle optimally, but that's fine.

This is another case of "good enough for where I am." Long-term, I think we should be sending Datadog's product logs to Datadog through a separate stream from user logs. Users have legitimately complained that sometimes dd-trace-php breaks their logs because of an unexpected format or bloat their log costs.

Reply:
Okay, still feels bit weird, but good enough :-)